### PR TITLE
[FLINK-28663][runtime] Allow multiple downstream consumer job vertices sharing the same intermediate dataset at scheduler side

### PR DESCRIPTION
**Pull request by wsry**
## What is the purpose of the change

Currently, one intermediate dataset can only be consumed by one downstream consumer vertex. If there are multiple consumer vertices consuming the same output of the same upstream vertex, multiple intermediate datasets will be produced. We can optimize this behavior to produce only one intermediate dataset which can be shared by multiple consumer vertices. As the first step, we should allow multiple downstream consumer job vertices sharing the same intermediate dataset at scheduler side. (Note that this optimization only works for blocking shuffle because pipelined shuffle result partition can not be consumed multiple times)

## Brief change log

  - Allow multiple downstream consumer job vertices sharing the same intermediate dataset at scheduler side

## Verifying this change

This change added tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptivebatch/forwardgroup/ForwardGroupComputeUtilTest.java [hunk 1]
```diff
@@ -83,14 +81,14 @@ public void testIsolatedVertices() throws Exception {
...
-    public void testVariousResultPartitionTypesBetweenVertices() throws Exception {
+    void testVariousResultPartitionTypesBetweenVertices() throws Exception {
         testThreeVerticesConnectSequentially(false, true, 1, 2);
         testThreeVerticesConnectSequentially(false, false, 0);
         testThreeVerticesConnectSequentially(true, true, 1, 3);
     }
 
     private void testThreeVerticesConnectSequentially(
-            boolean isForward1, boolean isForward2, int numOfGroups, int... groupSizes)
+            boolean isForward1, boolean isForward2, int numOfGroups, Integer... groupSizes)
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptivebatch/forwardgroup/ForwardGroupComputeUtilTest.java:91]: Why is this change needed?

**Review by zhuzhurk — COMMENTED** [review]
Thanks for opening this PR! @wsry 
I have just reviewed the test migration part and here are some comments for it.
Will continue reviewing the core changes commit.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobVertex.java [hunk 2]
```diff
@@ -73,7 +75,7 @@ public class JobVertex implements java.io.Serializable {
     private final List<OperatorIDPair> operatorIDs;
 
     /** List of produced data sets, one per writer. */
```
Comment (TanYuxin-tyx) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobVertex.java:77]: The annotation should also be modified.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartitionTest.java [hunk 3]
```diff
@@ -251,11 +279,24 @@ public static ExecutionGraph createExecutionGraph(
             v2.setMaxParallelism(consumerMaxParallelism);
         }
 
-        v2.connectNewDataSetAsInput(v1, distributionPattern, ResultPartitionType.BLOCKING);
+        final JobVertex v3 = new JobVertex("v3");
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartitionTest.java:280]: To double confirm, is this change helps to verify that the previous test can still pass with no exception with multi consumer job vertices?

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/RemoveCachedShuffleDescriptorTest.java [hunk 4]
```diff
@@ -294,42 +324,7 @@ private void testRemoveCacheForPointwiseEdgeAfterFailover(
         assertEquals(expectedAfter, blobWriter.numberOfBlobs());
     }
 
-    private DefaultScheduler createSchedulerAndDeploy(
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/RemoveCachedShuffleDescriptorTest.java:297]: It's better to keep this method for the convenience of tests in this test class. It can just invoke the `createSchedulerAndDeploy` in the util class though.

**Review by zhuzhurk — COMMENTED** [review]
Nice work! 
Here are a few comments.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartitionTest.java [hunk 5]
```diff
@@ -35,55 +35,48 @@
...
 
-import static org.hamcrest.CoreMatchers.is;
-import static org.hamcrest.MatcherAssert.assertThat;
-import static org.hamcrest.Matchers.equalTo;
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertTrue;
-
 /** Tests for {@link IntermediateResultPartition}. */
 public class IntermediateResultPartitionTest extends TestLogger {
```
[the remark this answers is not in the document]
Comment (wsry) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartitionTest.java:52]: 1. This test is used by another test, the ```public``` can not be removed.
2. Replace TestLogger with ```@ExtendWith({TestLoggerExtension.class})```

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptivebatch/forwardgroup/ForwardGroupComputeUtilTest.java (hunk 1 above)
Comment (wsry) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptivebatch/forwardgroup/ForwardGroupComputeUtilTest.java:91]: Because AssertJ's ```contains``` method does not accept int array.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartition.java [hunk 6]
```diff
@@ -47,6 +49,15 @@ public class IntermediateResultPartition {
     /** Whether this partition has produced some data. */
     private boolean hasDataProduced = false;
 
+    /**
+     * Released {@link ConsumedPartitionGroup}s for this result partition. This result partition can
+     * be released if all {@link ConsumedPartitionGroup}s are released.
+     */
+    private final Set<ConsumedPartitionGroup> releasedPartitionGroups = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (wsry) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartition.java:56]: 1. I use a hash set because it can tolerate duplicate releasing, though maybe there is no such case currently.
2. Renamed.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartitionTest.java (hunk 3 above)
Comment (wsry) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartitionTest.java:280]: Yes, exactly.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionGraphTestUtils.java [hunk 7]
```diff
@@ -528,4 +546,150 @@ public static void verifyGeneratedExecutionJobVertex(
...
+            Execution execution = vertex.getCurrentExecutionAttempt();
+            execution.registerProducedPartitions(slot.getTaskManagerLocation()).get();
+            execution.transitionState(ExecutionState.SCHEDULED);
+
+            vertex.tryAssignResource(slot);
+            vertex.deploy();
+        }
+    }
+
+    public static void transitionTasksToFinished(
```
[the remark this answers is not in the document]
Comment (wsry) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionGraphTestUtils.java:677]: 1. Moved and renamed the methods.
2. Use similar implementation will throw exception and fail the tests.

**Comment by wsry** [comment]
@zhuzhurk @TanYuxin-tyx Thanks for the review and feedbacks. I have updated the PR.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartitionTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartitionTest.java]: `@ExtendWith({TestLoggerExtension.class})` is not needed because it is by default activated for all tests.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adapter/DefaultExecutionTopologyTest.java [hunk 8]
```diff
@@ -191,18 +181,17 @@ public void testUpdateTopology() throws Exception {
 
         executionGraph.initializeJobVertex(ejv1, 0L);
         adapter.notifyExecutionGraphUpdated(executionGraph, Collections.singletonList(ejv1));
-        assertThat(IterableUtils.toStream(adapter.getVertices()).count(), is(3L));
+        Assertions.assertThat(IterableUtils.toStream(adapter.getVertices()).count()).isEqualTo(3L);
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adapter/DefaultExecutionTopologyTest.java]: Seems this comment is not fully addressed. There is not need to convert an `Iterable` to `Stream` to do `hasSize` check.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java [hunk 9]
```diff
@@ -718,31 +716,37 @@ public CompletableFuture<?> suspend() {
...
+
+                // ----------------------------------------------------------------
+                // Consumer is recovering or running => send update message now
+                // Consumer is deploying => cache the partition info which would be
+                // sent after switching to running
+                // ----------------------------------------------------------------
+                if ((consumerState == DEPLOYING
+                                || consumerState == RUNNING
+                                || consumerState == INITIALIZING)
+                        && !updatedVertices.contains(consumerVertexId)) {
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java:740]: Maybe do this check at the beginning of the inner loop? To avoid executing some unnecessary logics.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartition.java [hunk 10]
```diff
@@ -58,6 +66,25 @@ public IntermediateResultPartition(
         this.edgeManager = edgeManager;
     }
 
+    public void releaseConsumedPartitionGroup(ConsumedPartitionGroup partitionGroup) {
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartition.java:69]: -> markPartitionGroupReleasable()

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionGraphTestUtils.java [hunk 11]
```diff
@@ -249,6 +251,24 @@ public static void completeCancellingForAllVertices(ExecutionGraph eg) {
         }
     }
 
+    public static void finishJobVertex(ExecutionGraph executionGraph, JobVertexID jobVertexID) {
+
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionGraphTestUtils.java:255]: jobVertexID -> jobVertexId

I prefer to remove this empty line.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartitionTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (wsry) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartitionTest.java]: Removed ```@ExtendWith({TestLoggerExtension.class})```

**Comment by wsry** [comment]
@zhuzhurk I have updated the PR and add two fixup commits. Please take another look.

**Comment by wsry** [comment]
@flinkbot run azure

**Review by zhuzhurk — APPROVED** [review]
Thanks for addressing all the comments! @wsry 
The change looks good to me.
Would you squash it? I will merge it after CI gives green.
